File: takkernel/engine/src/main/cpp/renderer/feature/GLFeature.cpp

#include "feature/Feature.h"
#include "renderer/feature/GLFeature.h"
#include "renderer/feature/GLGeometry.h"
#include "renderer/feature/GLStyle.h"
#include "renderer/feature/GLStyleSpi.h"

//#include "renderer/feature/style/GLStyleSpi_CLI.h"

using namespace atakmap::feature;
using namespace atakmap::renderer;
using namespace atakmap::renderer::feature;
using namespace atakmap::renderer::map;

namespace {
    class FeatureRunnable {
        GLFeature *owner;
        PGSC::RefCountablePtr<atakmap::feature::Feature> feature;
    public:
        FeatureRunnable(const PGSC::RefCountablePtr<atakmap::feature::Feature> &feature, GLFeature *owner)
        : owner(owner), feature(feature) { }
        
        static void glRun(void *opaque) {
            FeatureRunnable *runnable = (FeatureRunnable *)opaque;
            PGSC::RefCountablePtr<atakmap::feature::Feature> &feature = runnable->feature;
            atakmap::renderer::feature::GLFeature *owner = runnable->owner;
            if (!Feature::isSame(*owner->subject, *feature)) {
                owner->subject = feature;
                owner->release();
            }
            delete runnable;
        }
    };
}
                
GLFeature::GLFeature(GLRenderContext *context, const PGSC::RefCountablePtr<atakmap::feature::Feature> &feature)
: context(context),
  subject(feature),
  initialized(false),
  style(nullptr),
  geometry(nullptr),
  styleCtx(nullptr) { }

void GLFeature::init(const GLMapView *view) {
    // XXX--get rid of the const_cast
    geometry = std::unique_ptr<GLGeometry>(GLGeometry::createRenderer(const_cast<Geometry *>(&subject->getGeometry())));
    
    GLStyleSpiArg arg;
    arg.style = subject->getStyle();
    arg.geometry = &subject->getGeometry();
    
    style = std::unique_ptr<GLStyle>(GLStyleFactory::create(arg));
    if (style != nullptr)
        this->styleCtx = style->createRenderContext(view, geometry.get());
    
    initialized = true;
}

Feature *GLFeature::getSubject() {
    return subject.get();
}

void GLFeature::draw(const GLMapView *view) {
    if (!initialized)
        init(view);
    if (style == nullptr)
        return;
    style->draw(view, geometry.get(), styleCtx);
}

void GLFeature::release()
{
    if (style != nullptr) {
        style->releaseRenderContext(styleCtx);
        styleCtx = nullptr;
        style = nullptr;
    }
    if (geometry != nullptr)
        geometry = nullptr;
    initialized = false;
}

bool GLFeature::isBatchable(const GLMapView *view) {
    if (!initialized)
        init(view);
    return (style == nullptr || style->isBatchable(view, geometry.get(), styleCtx));
}

void GLFeature::batch(const GLMapView *view, GLRenderBatch *batch) {
    if (!initialized)
        init(view);
    if (style == nullptr)
        return;
    style->batch(view, batch, geometry.get(), styleCtx);
}

void GLFeature::update(const PGSC::RefCountablePtr<atakmap::feature::Feature> &feature) {
    if (this->context->isGLThread() && !Feature::isSame(*subject, *feature)) {
        this->subject = feature;
        release();
    } else if (!context->isGLThread()) {
        // XXX - if feature IDs can be assumed to be the same, we can
        //       compare versions here as version must be monotonically
        //       increasing
        context->runOnGLThread(FeatureRunnable::glRun, new FeatureRunnable(feature, this));
    }
}

void GLFeature::start() { }

void GLFeature::stop()
{
    // XXX - 
}

File: takkernel/engine/src/main/java/com/atakmap/map/layer/feature/opengl/GLFeature.java
package com.atakmap.map.layer.feature.opengl;

import android.util.Pair;

import com.atakmap.annotations.DeprecatedApi;
import com.atakmap.map.MapRenderer;
import com.atakmap.map.layer.feature.Feature;
import com.atakmap.map.layer.feature.style.Style;
import com.atakmap.map.layer.feature.geometry.Geometry;
import com.atakmap.map.layer.feature.geometry.opengl.GLGeometry;
import com.atakmap.map.layer.feature.style.opengl.GLStyle;
import com.atakmap.map.layer.feature.style.opengl.GLStyleFactory;
import com.atakmap.map.opengl.GLMapBatchable;
import com.atakmap.map.opengl.GLMapRenderable;
import com.atakmap.map.opengl.GLMapSurface;
import com.atakmap.map.opengl.GLMapView;
import com.atakmap.opengl.GLRenderBatch;

/** @deprecated use the batch feature renderering framework */
@Deprecated
@DeprecatedApi(since = "4.1", forRemoval = true, removeAt = "4.4")
public final class GLFeature implements GLMapRenderable, GLMapBatchable {

    private MapRenderer surface;
    
    private Feature subject;
    private boolean initialized;

    private GLStyle style;
    private GLGeometry geometry;
    private GLStyle.StyleRenderContext styleCtx;
    
    public GLFeature(GLMapSurface surface, Feature feature) {
        this(surface.getGLMapView(), feature);
    }

    public GLFeature(MapRenderer surface, Feature feature) {
        this.surface = surface;
        this.subject = feature;
        this.styleCtx = null;
        
        this.initialized = false;
    }

    public Feature getSubject() {
        return this.subject;
    }
    
    public GLGeometry getGeometry(){
        if(this.geometry == null){
            this.geometry = GLGeometry.createRenderer(this.subject.getGeometry());
        }
        
        return this.geometry;
    }

    private void init(GLMapView view) {
        this.initialized = true;
        
        this.geometry = GLGeometry.createRenderer(this.subject.getGeometry());
        this.style = GLStyleFactory.create(Pair.<Style, Geometry>create(this.subject.getStyle(), this.subject.getGeometry()));
        if(this.style != null)
            this.styleCtx = this.style.createRenderContext(view, this.geometry);
    }

    public void update(final Feature feature) {
        if(GLMapSurface.isGLThread() && !Feature.isSame(this.subject, feature)) {
            this.subject = feature;
            this.release();
        } else if(!GLMapSurface.isGLThread()) {
            // XXX - if feature IDs can be assumed to be the same, we can
            //       compare versions here as version must be monotonically
            //       increasing
            this.surface.queueEvent(new Runnable() {
                @Override
                public void run() {
                    if(!Feature.isSame(GLFeature.this.subject, feature)) {
                        GLFeature.this.subject = feature;
                        GLFeature.this.release();
                    }
                }
            });
        }
    }

    @Override
    public boolean isBatchable(GLMapView view) {
        if(!this.initialized)
            this.init(view);
        return (this.style == null || this.style.isBatchable(view, this.geometry, this.styleCtx));
    }

    @Override
    public void batch(GLMapView view, GLRenderBatch batch) {
        if(!this.initialized)
            this.init(view);
        if(this.style == null)
            return;
        
        this.style.batch(view, batch, this.geometry, this.styleCtx);
    }

    @Override
    public void draw(GLMapView view) {
        if(!this.initialized)
            this.init(view);
        if(this.style == null)
            return;
        this.style.draw(view, this.geometry, this.styleCtx);
    }

    @Override
    public void release() {
        if(this.style != null) {
            this.style.releaseRenderContext(this.styleCtx);
            this.styleCtx = null;
            this.style = null;
        }
        if(this.geometry != null)
            this.geometry = null;
        this.initialized = false;
    }
}
